server/src/test/java/org/elasticsearch/action/admin/cluster/state/ClusterStateApiTests.java

```diff
@@ -0,0 +1,78 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.admin.cluster.state;
+
+import org.elasticsearch.action.ActionFuture;
+import org.elasticsearch.action.admin.cluster.settings.ClusterUpdateSettingsRequest;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+
+public class ClusterStateApiTests extends ESSingleNodeTestCase {
+
+    public void testWaitForMetaDataVersion() throws Exception {
+        ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+        ActionFuture<ClusterStateResponse> future1 = client().admin().cluster().state(clusterStateRequest);
+        assertThat(future1.isDone(), is(true));
+        assertThat(future1.actionGet().isWaitForTimedOut(), is(false));
+        long metadataVersion = future1.actionGet().getState().getMetaData().version();
+
+        // Verify that cluster state api returns after the cluster settings have been updated:
+        clusterStateRequest = new ClusterStateRequest();
+        clusterStateRequest.waitForMetaDataVersion(metadataVersion + 1);
+        clusterStateRequest.waitForTimeout(TimeValue.timeValueSeconds(1)); // Fail fast
+        ActionFuture<ClusterStateResponse> future2 = client().admin().cluster().state(clusterStateRequest);
+        assertThat(future2.isDone(), is(false));
+
+        ClusterUpdateSettingsRequest updateSettingsRequest = new ClusterUpdateSettingsRequest();
+        // Pick an arbitrary dynamic cluster setting and change it. Just to get metadata version incremented:
+        updateSettingsRequest.transientSettings(Settings.builder().put("cluster.max_shards_per_node", 999));
+        assertAcked(client().admin().cluster().updateSettings(updateSettingsRequest).actionGet());
+
+        assertBusy(() -> {
+            assertThat(future2.isDone(), is(true));
+        });
+        ClusterStateResponse response = future2.actionGet();
+        assertThat(response.isWaitForTimedOut(), is(false));
```

Comment:
On a really slow or overloaded machine, this can timeout in a second and cause spurious build failures. How about setting the timeout really high here (e.g., one hour). Then, in the test when you test that the timeout works, set the timeout low again.